Permit setting whether to use workload sets or loose manifests Fixes #34561

Forgind opened the pull request:
Fixes #34561

Adds a new subcommand (dotnet workload update --mode) taking one of three arguments: workloadset, loosemanifest, or auto. This will set the relevant bool in the install state according to that value.

That bool is not yet used.

Forgind committed:
Implement --mode

Forgind committed:
Add command option + tweak json

Forgind committed:
Change to a real Json parser

Forgind committed:
Include dotnetDir properly

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/MsiInstallerBase.cs:
```diff
@@ -497,9 +534,9 @@ protected void UpdateDependent(InstallRequestType requestType, string providerKe
         /// </summary>
         /// <param name="sdkFeatureBand">The feature band of the install state file.</param>
-        protected void RemoveInstallStateFile(SdkFeatureBand sdkFeatureBand)
+        protected void RemoveInstallStateFile(string dotnetDir, SdkFeatureBand sdkFeatureBand)
```
Great question!

How do we know if we're an MSI-based install or a file-based install? The current check goes to _dotnetDir_ to look for the msi sentinel file. That was the part that was actually throwing previously; as you said, we don't need to know dotnetDir if we're just writing to ProgramData, but we first check if we're an mis-based install, and that check threw an exception.

dsplaisted replied:
This is in the MsiInstallerBase class, so we know we're in an MSI-based install already.  We do need to pass the `dotnetDir` to `WorkloadInstallType.GetInstallStateFolder`, but rather than passing it through these methods, I think we can use the `DotNetHome` property, and if not we should be able to figure it out some other way.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/MsiInstallerBase.cs:
```diff
@@ -525,23 +562,23 @@ protected void RemoveInstallStateFile(SdkFeatureBand sdkFeatureBand)
         /// <param name="sdkFeatureBand">The path of the isntall state file to write.</param>
         /// <param name="jsonLines">The contents of the JSON file, formatted as a single line.</param>
-        protected void WriteInstallStateFile(SdkFeatureBand sdkFeatureBand, IEnumerable<string> jsonLines)
+        protected void WriteInstallStateFile(string dotnetDir, SdkFeatureBand sdkFeatureBand, Dictionary<string, string> jsonLines)
```
Naming: Since this method is no longer being passed all of the data that goes into the install state file, I think it might be clearer if it (and other similar methods) were renamed to something like `SaveInstallStateManifestVersions`.  The `jsonLines` parameter could be renamed to `manifestVersions`.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/IInstaller.cs:
```diff
@@ -38,14 +38,16 @@ internal interface IInstaller : IWorkloadManifestInstaller
-        void WriteInstallState(SdkFeatureBand sdkFeatureBand, IEnumerable<string> jsonLines);
+        void WriteInstallState(string dotnetDir, SdkFeatureBand sdkFeatureBand, Dictionary<string, string> jsonLines);
+
+        void AdjustInstallMode(string dotnetDir, SdkFeatureBand sdkFeatureBand, string newMode);
```
Nit: I think `UpdateInstallMode` would be more idiomatic than `AdjustInstallMode`.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerServer.cs:
```diff
@@ -94,15 +94,21 @@ public void Run()
 
+                        case InstallRequestType.AdjustWorkloadMode:
+                            AdjustInstallMode(request.DotnetDir, new SdkFeatureBand(request.SdkFeatureBand), request.ProductCode);
+                            string newMode = request.ProductCode.Equals("true", StringComparison.OrdinalIgnoreCase) ? "workload sets" : "loose manifests";
```
The PR description mentions 3 possible values: workloadset, loosemanifest, and auto.  It looks like this only handles the first two, and represents the value as a string.

Could we represent the value as an Enum, or possibly a nullable Boolean?

Forgind replied:
This doesn't handle 'auto' because auto is already handled in the WorkloadUpdateCommand where it is the same as loosemanifest (for now) but will later be changed to being the same as workloadset. Since there are only two possible values at this stage, I'm not sure I see the value of changing it to an enum.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs:
```diff
@@ -229,7 +229,11 @@ internal static void TryRunGarbageCollection(IInstaller workloadInstaller, IRepo
+                    }
+                    else
+                    {
+                        installer.DeleteInstallState(_dotnetPath, sdkFeatureBand);
```
Now that the install state file can include both the installation mode and the manifest versions, I don't think we want to delete the whole file here.  That would change the installation mode, which we don't want.  Rather, I think we want to remove the manifests from the install state file.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs:
```diff
@@ -80,6 +82,22 @@ public override int Execute()
+                }
+                else
+                {
+                    throw new GracefulException(string.Format(LocalizableStrings.WorkloadSetModeTakesWorkloadSetLooseManifestOrAuto, _workloadSetMode), isUserError: true);
```
I think if we use an Enum to represent the workload set mode (and use that enum in the option, ie `CliOption<WorkloadSetInstallMode>`), then System.CommandLine will handle this error checking for us.

Is this correct @baronfel?

dsplaisted reviewed (commented):
Looking good so far.

The mode isn't used yet, but we should probably have some tests that it is save and updated correctly.

Forgind committed:
Revert "Include dotnetDir properly"

This reverts commit b4ceac8abe3349391f3ecbfec2174acdd861089a.

Forgind committed:
Use DotNetHome instead of null

Forgind commented on src/Cli/dotnet/commands/dotnet-workload/install/MsiInstallerBase.cs:
```diff
@@ -272,6 +275,40 @@ protected uint RepairMsi(string productCode, string logFile)
+            }
+            else if (IsClient)
+            {
+                InstallResponseMessage response = Dispatcher.SendMsiRequest(InstallRequestType.AdjustWorkloadMode,
+                    null, null, newMode, dotnetDir, sdkFeatureBand.ToString());
```
I was actually thinking about going in the opposite direction and renaming this parameter something more generic like "argument"...every extra parameter is a small loss of performance, and I imagine we'll keep adding more things here until the requests are bulky and hard to use. I'd rather keep it slim and a bit generic. If you feel strongly, I can move towards new parameters instead.

dsplaisted replied:
I don't feel very strongly either way.  Thanks!

Forgind committed:
PR comments

Forgind committed:
Incorporate more feedback

Forgind committed:
Add test (and file that was missed?)

Forgind committed:
camelCase properties

This is to make them serialize into the install state file as camelCase, which is the default for json

dsplaisted commented on src/Cli/dotnet/Installer/Windows/InstallRequestMessage.cs:
```diff
@@ -120,6 +120,14 @@ public string WorkloadId
+        /// <summary>
+        /// The new mode to use: workloadset or loosemanifests
+        /// </summary>
+        public bool WorkloadMode
```
Naming nit: Calling a boolean property `WorkloadMode` doesn't tell you what true or false values mean.  I think it would be better as `UseWorkloadSets`, `WorkloadSetsEnabled`, or similar.  This also applies to method parameters where this value is passed, etc.

```suggestion
        public bool UseWorkloadSets
```

dsplaisted commented on src/Cli/dotnet/Installer/Windows/InstallRequestMessage.cs:
```diff
@@ -33,7 +33,7 @@ public string ManifestPath
         /// the JSON file to be written.
         /// </summary>
-        public string[] InstallStateContents
+        public Dictionary<string, string> InstallStateContents
```
Naming nit: I think `InstallStateManifestVersions` would be a bit clearer here.

```suggestion
        public Dictionary<string, string> InstallStateManifestVersions
```

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/InstallStateContents.cs:
```diff
@@ -0,0 +1,24 @@
+    internal class InstallStateContents
+    {
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public bool? useWorkloadSets { get; set; }
```
You can set up the Json serialization options to use camelCase, and then you can leave the C# properties PascalCase: https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/customize-properties?pivots=dotnet-7-0#use-a-built-in-naming-policy

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs:
```diff
@@ -452,21 +452,34 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
             if (File.Exists(path))
             {
-                File.Delete(path);
+                var installStateContents = InstallStateContents.FromString(File.Exists(path) ? File.ReadAllText(path) : "{}");
```
If the file doesn't exist, I think it's better to just create a new `InstallStateContents` object directly rather than parsing an empty JSON string.  See also other cases of this later in the file.

```suggestion
                var installStateContents = File.Exists(path) ? InstallStateContents.FromString(File.ReadAllText(path)) : new InstallStateContents();
```

dsplaisted commented on src/Cli/dotnet/commands/dotnet-tool/install/LocalizableStrings.resx:
```diff
@@ -170,6 +170,9 @@ Tool '{1}' (version '{2}') was successfully installed. Entry is added to the man
   <data name="LocalOptionDescription" xml:space="preserve">
     <value>Install the tool and add to the local tool manifest (default).</value>
   </data>
+  <data name="WorkloadSetMode" xml:space="preserve">
+    <value>Determine whether dotnet should use workload sets or loose manifests when installing or updating workloads.</value>
+  </data>
```
I think this string is in the wrong .resx file (this one is for tool installation, not workload installation).  It looks the string is also in the right .resx, so I think this can just be deleted.

dsplaisted commented on src/Cli/dotnet/commands/dotnet-workload/install/MsiInstallerBase.cs:
```diff
@@ -272,6 +275,40 @@ protected uint RepairMsi(string productCode, string logFile)
+                string path = Path.Combine(WorkloadInstallType.GetInstallStateFolder(sdkFeatureBand, DotNetHome), "default.json");
+                // Create the parent folder for the state file and set up all required ACLs
+                SecurityUtils.CreateSecureDirectory(Path.GetDirectoryName(path));
+                var installStateContents = InstallStateContents.FromString(File.Exists(path) ? File.ReadAllText(path) : "{}");
```
Same as the file based installer, I'd suggest creating an `InstallStateContents` object directly instead of parsing JSON if the file doesn't exist.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenFileBasedWorkloadInstall.cs:
```diff
@@ -26,6 +27,46 @@ public GivenFileBasedWorkloadInstall(ITestOutputHelper log) : base(log)
+            var sdkFeatureBand = new SdkFeatureBand(stringFeatureBand);
+            var path = Path.Combine(dotnetRoot, "metadata", "workloads", stringFeatureBand, "InstallState", "default.json");
+
+            // The install state shouldn't exist, but delete it in case it wasn't cleaned up properly from the last run.
+            if (File.Exists(path))
+            {
+                File.Delete(path);
+            }
```
This shouldn't be necessary.  The test infrastructure automatically makes sure that when it creates a test directory, that it's empty.  If the directory already exists, it will first delete it.  If that doesn't work, the test will fail.

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/GivenFileBasedWorkloadInstall.cs:
```diff
@@ -26,6 +27,46 @@ public GivenFileBasedWorkloadInstall(ITestOutputHelper log) : base(log)
+            installState.useWorkloadSets.Should().BeTrue();
+
+            installer.UpdateInstallMode(sdkFeatureBand, false);
+            installState = InstallStateContents.FromString(File.ReadAllText(path));
```
I would add a check here to be sure it still preserved the manifest versions.

```suggestion
            installState = InstallStateContents.FromString(File.ReadAllText(path));
            installState.manifests.Count.Should().Be(2);
```

dsplaisted commented on src/Tests/dotnet-workload-install.Tests/MockPackWorkloadInstaller.cs:
```diff
@@ -159,20 +165,24 @@ public void ReplaceWorkloadResolver(IWorkloadResolver workloadResolver)
             if (File.Exists(path))
             {
-                File.Delete(path);
+                var installStateContents = InstallStateContents.FromString(File.Exists(path) ? File.ReadAllText(path) : "{}");
```
More instances where we can do `new InstallStateContents()` instead of parsing `{}`.

dsplaisted reviewed (approved):
Looks good.

See my comments, but feel free to merge when ready.

Forgind committed:
Revert "camelCase properties"

This reverts commit 4f6b630f524777213bec042d7279bbb1d7bb1d3b.

Forgind committed:
PR comments

Forgind committed:
Use options universally
